### PR TITLE
Add IPv6 support and a new '-A' option to specify an IPv6 adress to bind

### LINKED ISSUE #41: HTTP frontend should support IPv6
Currently, http.c [only supports IPv4 addresses](https://github.com/urweb/urweb/blob/1a9520d6e00c63baca356d7e6aa2acc7424148bb/src/c/http.c#L361). You can trivially work around this using an IPv6-enabled reverse proxy, but native support is useful for development, and it’s easy to add.

### PR DESCRIPTION
**Pull request by sielenk**
This resolves #41, I think. At least the quick tests I did were successful.

### CODE REVIEW

**Comment by sielenk** [comment]
Hm, wierd. It worked on my machine...
"Listener socket bind failed" it says, so the bind call returned an error.
The API calls work and the IPv4 are successful, tough.
Is IPv6 not configured in the CI VMs?

**Comment by achlipala** [comment]
I'm not sure.  @JasonGross set them up, I believe, so maybe he can help.

**Comment by JasonGross** [comment]
Does something special have to be done to configure IPv6?  The machines are vanilla installs of Ubuntu, I think, but you can stick more commands in `.travis.yml` to set things up...

**Comment by JasonGross** [comment]
According to [this](https://docs.travis-ci.com/user/ci-environment/#Networking) and [this](https://github.com/travis-ci/travis-ci/issues/4964#issuecomment-159748141), it looks like IPv6 is unsupported on Travis.  I think running `make test` should not fail if the machine being run on does not support IPv6; maybe either check for IPv6 support before running the tests, or make a separate target for testing IPv6?

File: src/c/http.c [hunk 1]
```diff
@@ -413,8 +425,8 @@ int main(int argc, char *argv[]) {
     return 1;
   }
```
Comment (jmitchell) [inline src/c/http.c]: Consider adding another `setsockopt` that disables `IPV6_V6ONLY` to explicitly allow IPv4-to-IPv6 mapped addresses. Not all operating systems use the same default value.

Comment (sielenk) [inline src/c/http.c]: Agreed, and on Linux it can have a different default value per system.

**Comment by sielenk** [comment]
@JasonGross: IPv6 can not be entirely unsupported in the VMs since the code uses the IPv6 API and connecting via the IPv4 localhost works.
I assume there is just no interface configured with an IPv6 address. For the tests to run, only 'lo' would need one ('::1' to be precise). External connections via other interfaces are irrelevant.

File: Makefile.am [hunk 2]
```diff
@@ -116,9 +116,22 @@ test:
 	bin/urweb -boot -noEmacs -dbms sqlite -db $(TESTDB) -demo /Demo demo
 	rm -f $(TESTDB)
 	sqlite3 $(TESTDB) < demo/demo.sql
-	demo/demo.exe & echo $$! > $(TESTPID)
+	demo/demo.exe -a 127.0.0.1 & echo $$! > $(TESTPID)
```
Comment (JasonGross) [inline Makefile.am]: Should there also be a test that not passing any port works (and for CI, works even without IPv6 sort)?

**Comment by JasonGross** [comment]
@sielenk quoting https://github.com/travis-ci/travis-ci/issues/4964#issuecomment-159748141 :
> I understanding this may be disappointing, but I can tell you that having IPv6 has been a by-product of where we've been hosting precise before, but it's not something we consider a core feature that we support in Travis CI at this time and it will go away as we migrate away from that hosting platform.
>
> Unfortunately, the major public cloud providers do not provide any supported methods for having internal IPv6 connectivity nor are there any real options for in-bound IPv6 external connectivity in a way that is usable for Travis's build environments, so we're limited by what is supported on the providers which give us the best economies of scale and overall feature sets we need.

At this point, you know basically as much, or more, than I do.

File: Makefile.am (hunk 2 above)
Comment (sielenk) [inline Makefile.am]: You mean we should keep the original one and add my change as additional test?
I think then I would like to reorganize the code a little; having almost the same code thrice is a little too much for my taste...

**Comment by achlipala** [comment]
FWIW, that `test` target was never meant to be exhaustive, so just removing new IPv6-related tests would make sense to me.

**Comment by achlipala** [comment]
Well, I'm happy to bless reasonable-looking solutions, but given my general ignorance of IPv6 and Travis, I'll wait for what looks like a conclusive statement by someone else here.

**Comment by jmitchell** [comment]
The IPv6 end of things looks good as far as I can tell. In addition to reviewing the code I manually confirmed a simple project accepts connections via `http://127.0.0.1:<port>` and `http://[::1]:<port>`.

**Pull request merged**